consult.el

```diff
@@ -1044,6 +1044,55 @@ The arguments and expected return value are as specified for
         (funcall exit completion exit-status))
       t)))
 
+(defun consult--extract-commands (mode)
+  "Extract commands from MODE."
+  (let ((library-path (symbol-file mode))
+        (mode-rx (regexp-quote
+                  (substring (if (eq major-mode 'c-mode)
+                                 "cc-mode"
+                               (symbol-name major-mode))
+                             0 -5))))
+    (mapcan
+     (lambda (feature)
+       (let ((path (car feature)))
+         (when (or (string= path library-path)
+                   (string-match-p mode-rx (file-name-nondirectory path)))
+           (mapcar #'cdr
+                   (seq-filter
+                    (lambda (item)
+                      (and (consp item)
+                           (eq (car item) 'defun)
+                           (commandp (cdr item))
+                           (not (string-match-p "--" (symbol-name cmd)))))
+                    (cdr feature))))))
+     load-history)))
+
+;;;###autoload
+(defun consult-major-command ()
+  "Run a command from the current major mode."
+  (interactive)
+  (call-interactively
+   (consult--read "M-X: "
+                  (consult--extract-commands major-mode)
```

Comment:
Then the question is if you also want to include the local key bindings. This is what the smex version does, but maybe it doesn't add value?

Reply:
I didn't include them because @protesilaos didn't. I would expect all of the commands from key bindings to already be included, but I'm not sure.

Reply:
Ok, we can quickly check this, by copying over the snippet from my branch. But unfortunately it wouldn't work well together with the narrowing idea since from the keymap the provenance is unclear.

Reply:
I decided to not add the binding stuff. Keeping separate concepts separate.